jimfs/src/main/java/com/google/common/jimfs/PollingWatchService.java

```diff
@@ -0,0 +1,250 @@
+
+  private boolean isSameFileSystem(Path path) {
+    return ((JimfsFileSystem) path.getFileSystem()).getDefaultView() == view;
+  }
+
+  @VisibleForTesting
+  synchronized boolean isPolling() {
+    return pollingFuture != null;
+  }
+
+  @Override
+  public synchronized void cancelled(Key key) {
+    snapshots.remove(key);
+
+    if (snapshots.isEmpty()) {
+      stopPolling();
+    }
+  }
+
+  @Override
+  public void close() {
+    super.close();
+
+    synchronized (this) {
+      // synchronize to ensure no new
+      for (Key key : snapshots.keySet()) {
+        key.cancel();
+      }
+
+      pollingService.shutdown();
+      fileSystemState.unregister(this);
+    }
+  }
+
+  private void startPolling() {
+    pollingFuture = pollingService.scheduleAtFixedRate(pollingTask, interval, interval, timeUnit);
+  }
+
+  private void stopPolling() {
+    pollingFuture.cancel(false);
+    pollingFuture = null;
+  }
+
+  private final Runnable pollingTask =
+      new Runnable() {
+        @Override
+        public void run() {
+          synchronized (PollingWatchService.this) {
+            for (Map.Entry<Key, Snapshot> entry : snapshots.entrySet()) {
+              Key key = entry.getKey();
+              Snapshot previousSnapshot = entry.getValue();
+
+              JimfsPath path = (JimfsPath) key.watchable();
+              try {
+                Snapshot newSnapshot = takeSnapshot(path);
+                boolean posted = previousSnapshot.postChanges(newSnapshot, key);
+                entry.setValue(newSnapshot);
+                if (posted) {
+                  key.signal();
+                }
+              } catch (IOException e) {
+                // snapshot failed; assume file does not exist or isn't a directory
+                // and cancel the key
+                key.cancel();
+              }
+            }
+          }
+        }
+      };
+
+  private Snapshot takeSnapshot(JimfsPath path) throws IOException {
+    return new Snapshot(view.snapshotModifiedTimes(path));
+  }
+
+  /** Snapshot of the state of a directory at a particular moment. */
+  private final class Snapshot {
+
+    /** Maps directory entry names to last modified times. */
+    private final ImmutableMap<Name, Long> modifiedTimes;
+
+    Snapshot(Map<Name, Long> modifiedTimes) {
+      this.modifiedTimes = ImmutableMap.copyOf(modifiedTimes);
+    }
+
+    /**
+     * Posts events to the given key based on the kinds of events it subscribes to and what events
+     * have occurred between this state and the given new state.
+     */
+    boolean postChanges(Snapshot newState, Key key) {
+      boolean changesPosted = false;
+
+      if (key.subscribesTo(ENTRY_CREATE)) {
+        Set<Name> created =
+            Sets.difference(newState.modifiedTimes.keySet(), modifiedTimes.keySet());
+
+        for (Name name : created) {
+          key.post(new Event<>(ENTRY_CREATE, 1, pathService.createFileName(name)));
+          changesPosted = true;
+        }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 214:222, 224:232.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.